### PR TITLE
Optimize Butina Clustering for Performance and Expand API

### PR DESCRIPTION
**Pull request by evasnow1992**
#### What does this implement/fix? Explain your changes.

- Optimized Butina clustering algorithm using NumPy arrays for improved performance
- Enhanced modularity and overall code readability
- Expanded API to support direct use of NumPy arrays as input
- Added support for nxn input distance matrices

### CODE REVIEW

**Comment by evasnow1992** [comment]
We've benchmarked the performance of our optimized Butina clustering implementation against the current RDKit implementation. The first figure below shows the times of speed up of clustering 10,000 randomly selected molecules from ZINC15 dataset in SMILES using Morgan fingerprints for distance computation with a distance threshold of 0.7. For our implementation, when using a list of data as input, it consistently achieves ~3x speedup compared to the current RDKit implementation. If allowing using a NumPy nxn distance matrix, the speedup can be as high as 10x and is likely to increase as the number of molecules increases.

We understand that the actual speedup may be affected by data density or the selection of distance threshold. So in the second figure below we show the speedup of our implementation, with the list and NumPy inputs for different numbers of resulting clusters. When using list input, our implementation achieves 1-5x speedup across different data densities. If a NumPy distance matrix is used, the speedup can be as high as 100x for sparse data (large cluster number).

Picture1
Picture2

File: rdkit/ML/Cluster/Butina.py [hunk 1]
```diff
@@ -59,57 +65,83 @@ def ClusterData(data, nPts, distThresh, isDistData=False, distFunc=EuclideanDist
...
-  tLists = [(len(y), x) for x, y in enumerate(nbrLists)]
-  tLists.sort(reverse=True)
-
-  res = []
-  seen = [0] * nPts
-  while tLists:
-    _, idx = tLists.pop(0)
+  if isDistData:
+    # Check if data is a supported type
+    if not isinstance(data, (list, np.ndarray)):
```
Comment (greglandrum) [inline rdkit/ML/Cluster/Butina.py:70]: Any reason to not also support tuples here?

File: rdkit/ML/Cluster/Butina.py [hunk 2]
```diff
@@ -59,57 +65,83 @@ def ClusterData(data, nPts, distThresh, isDistData=False, distFunc=EuclideanDist
...
+      raise ValueError(f"Unsupported type for data, {type(data)}")
+    
+    # Check if data is a 1D array or list
+    if isinstance(data, list) or isinstance(data, np.ndarray) and data.ndim == 1:
```
Comment (greglandrum) [inline rdkit/ML/Cluster/Butina.py:74]: readability suggestion:
```suggestion
    if isinstance(data, list) or (isinstance(data, np.ndarray) and data.ndim == 1):
```

File: rdkit/ML/Cluster/Butina.py [hunk 3]
```diff
@@ -59,57 +65,83 @@ def ClusterData(data, nPts, distThresh, isDistData=False, distFunc=EuclideanDist
...
-  tLists.sort(reverse=True)
-
-  res = []
-  seen = [0] * nPts
-  while tLists:
-    _, idx = tLists.pop(0)
+  if isDistData:
+    # Check if data is a supported type
+    if not isinstance(data, (list, np.ndarray)):
+      raise ValueError(f"Unsupported type for data, {type(data)}")
```
Comment (greglandrum) [inline rdkit/ML/Cluster/Butina.py:71]: If the argument type is wrong, shouldn't this be a `TypeError`?

File: rdkit/ML/Cluster/UnitTestButina.py [hunk 4]
```diff
@@ -1,6 +1,6 @@
 # $Id$
```
Comment (greglandrum) [inline rdkit/ML/Cluster/UnitTestButina.py:1]: please remove this SVN dropping too.

**Review by greglandrum — CHANGES_REQUESTED** [review]
The performance data looks great! Thanks for doing this.

I made a few minor cleanup suggestions.

**Comment by evasnow1992** [comment]
Hi Greg, thank you for your feedback. We've updated the scripts to address the issues you raised, including cleaning up stale comments, improving input data type handling, and updating the test expectation in UnitTestButina.py. We've pushed the updated changes to this PR. Please let us know if you have any further suggestions or revisions.

**Comment by greglandrum** [comment]
@evasnow1992 the changes look good, thanks.
I see that you've marked this as draft. do you plan further changes?

**Comment by evasnow1992** [comment]
Hi Greg, thank you for the quick reply. I had marked this as a draft as I was waiting for all the tests to complete. No further changes are planned. Please let me know if you feel it's ready for merging or any further revision is needed.

**Review by greglandrum — APPROVED** [review]
LGTM

**Comment by greglandrum** [comment]
Thanks for the contribution @evasnow1992 !

**Pull request merged**